src/services/import-service.js

```diff
@@ -59,6 +74,68 @@ class ImportService {
 
     return importableDocuments;
   }
+
+  async createImportBatch({ importSource, documentsToImport, user }) {
+    const batch = {
+      _id: uniqueId.create(),
+      createdBy: user._id,
+      createdOn: new Date(),
+      completedOn: null,
+      batchType: BATCH_TYPE.importDocuments,
+      batchParams: {
+        source: importSource.name
+      },
+      errors: []
+    };
+
+    const tasks = documentsToImport.map(doc => ({
+      _id: uniqueId.create(),
+      batchId: batch._id,
+      taskType: TASK_TYPE.importDocument,
+      processed: false,
+      attempts: [],
+      taskParams: {
+        key: doc.key,
+        title: doc.title,
+        slug: doc.slug,
+        language: doc.language,
+        updatedOn: new Date(doc.updatedOn),
+        importedRevision: doc.importedRevision,
+        importableRevision: doc.importableRevision,
+        importType: doc.importType
+      }
+    }));
+
+    let lock;
+    try {
+      lock = await this.batchLockStore.takeLock(importSource.name);
+    } catch (error) {
+      throw new BadRequest('Concurrent batch creation for the same source is not allowed');
+    }
+
+    try {
+      const existingActiveBatch = await this.batchStore.findOne({
+        'batchType': BATCH_TYPE.importDocuments,
+        'batchParams.source': importSource.name,
+        'completedOn': null
+      });
+
+      if (existingActiveBatch) {
+        throw new BadRequest('Cannot create a new batch while another batch for the same source is still active');
```

Comment:
Tiny thing, should we extract a `throwConcurrentBatchError` and have the same message in both cases (lines 113 and 124) ?

Reply:
They're not exactly the same thing, but the second message kind of includes the first case, so yeah, I'll do that

Reply:
I know, there's a slight difference but boils down to the same restriction.

Reply:
I extracted the error message into a constant and made the test check for the error type instead of the message

Reply:
I didn't want to extract a function in order to see the correct line number in stack traces on top where the error happened

Reply:
Genius 🧠 Great!